README.md
# Nano-IoT-Environment-Logger

Using the Arduino Nano IoT 33 and either a BME280, SHTC3, or ADT7410 sensor send environmental
measurements to an MQTT broker.

## Features

### Multiple Sensors

The program supports one of several sensors.

- BME280 - temperature, humidity, and pressure
- SHTC3 - temperatue, humididy
- ADT7410 - temperature

### BLE Configuration

During the first minute after reset connect to the board with BLE to change the default configuration.
The configuration will be saved in flash memory.

Note, you must always set the WiFi password when updating the configuration.

### Battery Monitor

If enabled the charge level of the battery will be included in the MQTT message.

### LUA Script for BlueSee

Use BlueSee with the included LUA script for a simple UI to configure the board via BLE.


History.md
# History

## 26 Nov 2021

Working on the stability of the code.  Infrequently, every 1 or 2 days, the system will reboot due to watchdog timeout.  While I've not yet spotted it happening live in the logs with a terminal attached I believe what is happening is that after a long delay in establishing the wifi connection if the send to the MQTT broker is also slow it could extend beyond the length of time of the watchdog.  In connect Wifi I was not resetting the watchdog timer after the system eventually connected.  Also in this section of the code I decreased the retries, as with 10 it was getting close to the watchdog timeout limit.  Additionally after timing the loop I did see instances where the time to send the message was over 2 times as long as average.  This also could get the near to the watchdog timeout limit.  Therefore I added Watchdog resets right before and after sending to MQTT.


ConfigService.h
#ifndef CONFIGSERVICE_H
#define CONFIGSERVICE_H

#include <ArduinoBLE.h>

/**
 * @brief Provide a BLE service to allow reading and updating the
 * configuration information for the application.  Once
 * locked the settings are retained in flash memory.
 */
class ConfigService {
public:
  ConfigService(const char *p_ssid, const char *p_passwd,
                const char *p_hostName, const char *p_location, 
                const char *p_mqttBroker, const char *p_topicRoot,
                unsigned int p_sampleInterval);

  /** @brief Initialize the service, call before use. */
  void begin();

  /** @return true if initialized */
  bool initialized() const;

  /** @return a reference to the ConfigService */
  BLEService &getConfigService();

  void debug_print_configuration();
  
  char ssid[64];         /** the ssid for WiFi */
  char wifiPassword[64]; /** the wifi Password */
  char hostName[32];     /** host name used for as the local device name */
  char mqttBroker[64];   /** FQDN address of the mqtt server */
  char topicRoot[64];    /** mqtt topic name to use as the base name */
  char location[64];     /** sensor location */
  unsigned int sampleInterval; /** not currently used! */

  bool isInitialized; /** True if initialized from Flash memory */
  bool isLocked;      /** True if the configuration has been locked */

  /** @brief process callbacks for the the service characteristics */
  friend void onLock(BLEDevice central, BLECharacteristic characteristic);

protected:
  BLEService service_;
  BLEStringCharacteristic ssidCharacteristic_;
  BLEDescriptor ssidDescriptor_;
  BLEStringCharacteristic wifiPasswordCharacteristic_;
  BLEDescriptor wifiPasswordDescriptor_;
  BLEStringCharacteristic hostNameCharacteristic_;
  BLEDescriptor hostNameDescriptor_;
  BLEStringCharacteristic mqttBrokerCharacteristic_;
  BLEDescriptor mqttBrokerDescriptor_;
  BLEStringCharacteristic topicRootCharacteristic_;
  BLEDescriptor topicRootDescriptor_;
  BLEStringCharacteristic locationCharacteristic_;
  BLEDescriptor locationDescriptor_;
  BLEUnsignedIntCharacteristic sampleIntervalCharacteristic_;
  BLEUnsignedIntCharacteristic isLockedCharacteristic_;
  BLEDescriptor sampleIntervalDescriptor_;
  BLEStringCharacteristic lockCharacteristic_;

private:
  char configurationPassword[64]; /** password to lock and unlock the
                                   configuration */
};

#endif


MovingAverageFilter.h
/**
 * Simple moving average filter.  It takes template parameters of the data type
 * as well as the Number of Taps (or size of the buffer).
 */

template <class DataType, unsigned int NbrOfTaps>
class MovingAverageFilter {
  
  public:
    MovingAverageFilter():
      bufferIndex(0)
      {
        memset(buffer, 0, sizeof(buffer));
      }

    /** @brief Add a sample to the filer buffer */
    void addSample(DataType s) {
      buffer[bufferIndex] = s;
      bufferIndex = (bufferIndex + 1) % NbrOfTaps;
    }

    /** @brief Compute the current moving average across the buffer */
    const DataType value() const {
      DataType sum = 0;
      for (auto i = 0; i < NbrOfTaps; i++) {
        sum += buffer[i];
      }
      return sum / NbrOfTaps;
    }

    /** @return The length of the sample buffer */
    unsigned int nbrOfTaps() {
      return NbrOfTaps;
    }

  private:
    DataType buffer[NbrOfTaps];
    unsigned int bufferIndex;
};

IoTCentral-BlueSee.lua
local bluesee = require "bluesee"

-- Config service definition
local config_service_uuid = bluesee.UUID.new('44af2abf-f3d9-4429-bbb8-ec770f1e355a')
bluesee.set_display_name(config_service_uuid, 'IoTCentral Configuration')
bluesee.set_display_category(config_service_uuid, bluesee.ui)

local wifiSSID_uuid = bluesee.UUID.new('0465ac13-a0f3-46a6-990a-5a04b32b3b60')
local wifiPassword_uuid = bluesee.UUID.new('27030384-eac9-4907-8e44-5c16d778aa7a')
local hostname_uuid = bluesee.UUID.new('ef42bf97-1b9c-4d45-941d-d60dc564dc6f')
local mqttBroker_uuid = bluesee.UUID.new('0d071785-b22b-49d6-86be-270de52da930')
local topicRoot_uuid = bluesee.UUID.new('18cda5b0-3b76-4319-9716-acd1a409d3f6')
local location_uuid = bluesee.UUID.new('69394063-f713-4ff7-be75-77d614d91709')
local sampleInterval_uuid = bluesee.UUID.new('1682229f-bb5c-4f4a-96a9-1027f13d83f9')
local isConfigurationLocked_uuid = bluesee.UUID.new('d02db20e-6e1f-4541-bd3d-7715e00b2b82')
local configurationLock_uuid = bluesee.UUID.new('29636a43-d59a-46a1-ad0a-34aa23a0e90c')
local configurationUnlock_uuid = bluesee.UUID.new('508458ce-2dad-42de-aad1-4969608adb6a')

-- Create a collection of controls to process
-- readable and/or writeable text strings
-- If the characteristic is readable a label
-- is included whos value is the characteristic
-- If the characteristic is writable a textfield
-- control is included and the write button will
-- update the characteristic
function add_ReadWriteTextControl(span, title, ch)

    local label = nil
    local textfield = nil
    local read_btn = nil
    local write_btn = nil

    if ch.readable then
        label = bluesee.new_widget(bluesee.label)
        label.title = title
        read_btn = bluesee.new_widget(bluesee.button)
        read_btn.title = "Read"
        read_btn.on_click = function()
            ch:read()
        end
    end

    if ch.writeable then
        textfield = bluesee.new_widget(bluesee.textfield)
        if ch.readable then
            textfield.title = "New Value"
        else
            textfield.title = title
        end
        write_btn = bluesee.new_widget(bluesee.button)
        write_btn.title = "Write"
        write_btn.on_click = function()
            ch:write_binary(textfield.value)
            if ch.readable then
                ch:read()   -- force a read to refresh the label
            end
        end
    end

    span:add_widget(bluesee.new_widget(bluesee.hr))
    if label ~= nil then
        span:add_widget(label)
    end
    if read_btn ~= nil then
        span:add_widget(read_btn)
    end
    if textfield ~= nil then
        span:add_widget(textfield)
    end
    if write_btn ~= nil then
        span:add_widget(write_btn)
    end

    if ch.readable then 
        local update_fn = function()
            local val = tostring(ch.value)
            if val == nil then
                val = ch.value:as_hex_string()
            end
            label.value = val
        end
        ch:add_read_callback(update_fn)
        ch:read()
    end

end

function add_ReadWriteUnsignedIntegerControl(span, title, ch)

    local label = nil
    local textfield = nil
    local read_btn = nil
    local write_btn = nil

    if ch.readable then
        label = bluesee.new_widget(bluesee.label)
        label.title = title
        read_btn = bluesee.new_widget(bluesee.button)
        read_btn.title = "Read"
        read_btn.on_click = function()
            ch:read()
        end
    end

    if ch.writeable then
        textfield = bluesee.new_widget(bluesee.textfield)
        if ch.readable then
            textfield.title = "New Value"
        else
            textfield.title = title
        end
        write_btn = bluesee.new_widget(bluesee.button)
        write_btn.title = "Write"
        write_btn.on_click = function()
            local val = tonumber(textfield.value)
            ch:write(bluesee.Data.new(val, 4, bluesee.Data.endian_little))
            if ch.readable then
                ch:read()   -- force a read to refresh the label
            end
        end
    end

    span:add_widget(bluesee.new_widget(bluesee.hr))
    if label ~= nil then
        span:add_widget(label)
    end
    if read_btn ~= nil then
        span:add_widget(read_btn)
    end
    if textfield ~= nil then
        span:add_widget(textfield)
    end
    if write_btn ~= nil then
        span:add_widget(write_btn)
    end

    if ch.readable then 
        local update_fn = function()
            local val = tostring(ch.value:unsigned_integer(bluesee.Data.endian_little))
            if val == nil then
                val = ch.value:as_hex_string()
            end
            label.value = val
        end
        ch:add_read_callback(update_fn)
        if ch.subscribable then
            ch:subscribe(update_fn)
        end
        ch:read()
    end

end

-- Register the Config service
bluesee.register_service(config_service_uuid, function(span)

    
    -- Build the UI and add handlers based on the available and known
    -- characteristics
    span.on_ch_discovered = function(ch)

        if ch.uuid == wifiSSID_uuid then
            add_ReadWriteTextControl(span, "SSID", ch)
        elseif ch.uuid == wifiPassword_uuid then
            add_ReadWriteTextControl(span, "WiFi Password", ch)
        elseif ch.uuid == hostname_uuid then
            add_ReadWriteTextControl(span, "Hostname", ch)
        elseif ch.uuid == mqttBroker_uuid then
            add_ReadWriteTextControl(span, "MQTT Broker", ch)
        elseif ch.uuid == topicRoot_uuid then
            add_ReadWriteTextControl(span, "Root Topic", ch)
        elseif ch.uuid == location_uuid then
            add_ReadWriteTextControl(span, "Location", ch)
        elseif ch.uuid == sampleInterval_uuid then
            add_ReadWriteUnsignedIntegerControl(span, "Sample Interval", ch)
        elseif ch.uuid == isConfigurationLocked_uuid then
            add_ReadWriteUnsignedIntegerControl(span, "Configuration Is Locked", ch)
        elseif ch.uuid == configurationLock_uuid then
            add_ReadWriteTextControl(span, "Lock/Unlock Configuration", ch)
        else
            span:log("Error - unknown characteristic" .. ch.uuid)
        end
    end

 end)

ConfigService.cpp
#include "ConfigService.h"

// #define _DEBUG_
#include "Debug.h"

#include <FlashAsEEPROM.h>

typedef struct {
  int valid;
  char ssid[64];
  char wifiPassword[64];
  char hostName[32];
  char mqttBroker[64];
  char topicRoot[64];
  char location[64];
  unsigned int sampleInterval;
  char configurationPassword[64];
} ConfigServiceEEPROM_t;

FlashStorage(flash_configuration, ConfigServiceEEPROM_t);

static ConfigService
    *singleton; /** the one and only instance, used in the callbacks */

void onLock(BLEDevice central, BLECharacteristic characteristic);

ConfigService::ConfigService(const char *p_ssid, const char *p_passwd,
                             const char *p_hostName, const char *p_location,
                             const char *p_mqttBroker,
                             const char *p_topicRoot,
                             unsigned int p_sampleInterval)
    : service_("44af2abf-f3d9-4429-bbb8-ec770f1e355a"),
      ssidCharacteristic_("0465ac13-a0f3-46a6-990a-5a04b32b3b60",
                          BLERead | BLEWrite, sizeof(ssid)),
      ssidDescriptor_("2901", "WiFi SSID"),
      wifiPasswordCharacteristic_("27030384-eac9-4907-8e44-5c16d778aa7a",
                                  BLEWrite, sizeof(wifiPassword)),
      wifiPasswordDescriptor_("2901", "WiFi Password"),
      hostNameCharacteristic_("ef42bf97-1b9c-4d45-941d-d60dc564dc6f",
                              BLERead | BLEWrite, sizeof(hostName)),
      hostNameDescriptor_("2901", "Host Name"),
      mqttBrokerCharacteristic_("0d071785-b22b-49d6-86be-270de52da930",
                                BLERead | BLEWrite, sizeof(mqttBroker)),
      mqttBrokerDescriptor_("2901", "MQTT Address"),
      topicRootCharacteristic_("18cda5b0-3b76-4319-9716-acd1a409d3f6",
                               BLERead | BLEWrite, sizeof(topicRoot)),
      topicRootDescriptor_("2901", "MQTT Root Topic"),
      locationCharacteristic_("69394063-f713-4ff7-be75-77d614d91709", 
                               BLERead | BLEWrite, sizeof(location)),
      locationDescriptor_("2901", "Sensor Location"),
      sampleIntervalCharacteristic_("1682229f-bb5c-4f4a-96a9-1027f13d83f9",
                                    BLERead | BLEWrite),
      sampleIntervalDescriptor_("2901", "Sample Interval in Seconds"),
      isLockedCharacteristic_("d02db20e-6e1f-4541-bd3d-7715e00b2b82",
                              BLERead | BLENotify),
      lockCharacteristic_("29636a43-d59a-46a1-ad0a-34aa23a0e90c", BLEWrite,
                          sizeof(configurationPassword)) {
  strncpy(ssid, p_ssid, sizeof(ssid));
  strncpy(wifiPassword, p_passwd, sizeof(wifiPassword));
  strncpy(hostName, p_hostName, sizeof(hostName));
  strncpy(mqttBroker, p_mqttBroker, sizeof(mqttBroker));
  strncpy(topicRoot, p_topicRoot, sizeof(topicRoot));
  strncpy(location, p_location, sizeof(location));
  sampleInterval = p_sampleInterval;

  isInitialized = false;
  memset(configurationPassword, 0, sizeof(configurationPassword));
  isLocked = false;

  ConfigServiceEEPROM_t flashConfig;

  flashConfig = flash_configuration.read();
  if (flashConfig.valid == 99) {
    // Load the config settings from flash
    strncpy(ssid, flashConfig.ssid, sizeof(ssid));
    strncpy(wifiPassword, flashConfig.wifiPassword, sizeof(wifiPassword));
    strncpy(hostName, flashConfig.hostName, sizeof(hostName));
    strncpy(mqttBroker, flashConfig.mqttBroker, sizeof(mqttBroker));
    strncpy(topicRoot, flashConfig.topicRoot, sizeof(topicRoot));
    strncpy(location, flashConfig.location, sizeof(location));
    sampleInterval = flashConfig.sampleInterval;
    strncpy(configurationPassword, flashConfig.configurationPassword,
            sizeof(configurationPassword));

    isInitialized = true;
    isLocked = true;
  }
}

BLEService &ConfigService::getConfigService() { return service_; }

void ConfigService::begin() {

  singleton = this;

  ssidCharacteristic_.addDescriptor(ssidDescriptor_);
  wifiPasswordCharacteristic_.addDescriptor(wifiPasswordDescriptor_);
  hostNameCharacteristic_.addDescriptor(hostNameDescriptor_);
  mqttBrokerCharacteristic_.addDescriptor(mqttBrokerDescriptor_);
  topicRootCharacteristic_.addDescriptor(topicRootDescriptor_);
  locationCharacteristic_.addDescriptor(locationDescriptor_);
  sampleIntervalCharacteristic_.addDescriptor(sampleIntervalDescriptor_);

  // set the initial values
  ssidCharacteristic_.writeValue(ssid);
  hostNameCharacteristic_.writeValue(hostName);
  mqttBrokerCharacteristic_.writeValue(mqttBroker);
  topicRootCharacteristic_.writeValue(topicRoot);
  locationCharacteristic_.writeValue(location);
  sampleIntervalCharacteristic_.writeValue(sampleInterval);
  isLockedCharacteristic_.writeValue(isLocked ? 1 : 0);

  // Callbacks for lock and unlock
  lockCharacteristic_.setEventHandler(BLEWritten, onLock);

  service_.addCharacteristic(ssidCharacteristic_);
  service_.addCharacteristic(wifiPasswordCharacteristic_);
  service_.addCharacteristic(hostNameCharacteristic_);
  service_.addCharacteristic(mqttBrokerCharacteristic_);
  service_.addCharacteristic(topicRootCharacteristic_);
  service_.addCharacteristic(locationCharacteristic_);
  service_.addCharacteristic(sampleIntervalCharacteristic_);
  service_.addCharacteristic(isLockedCharacteristic_);
  service_.addCharacteristic(lockCharacteristic_);

  BLE.addService(service_);
}

void onLock(BLEDevice central, BLECharacteristic characteristic) {
  if (!singleton->isLocked) {
    DEBUG_PRINTF("Configuration Locked with password = %s\n",
                 singleton->lockCharacteristic_.value().c_str());
    singleton->isInitialized = true;
    singleton->isLocked = true;
    strncpy(singleton->configurationPassword,
            singleton->lockCharacteristic_.value().c_str(),
            sizeof(singleton->configurationPassword));
    strncpy(singleton->ssid, singleton->ssidCharacteristic_.value().c_str(),
            sizeof(singleton->ssid));
    strncpy(singleton->wifiPassword,
            singleton->wifiPasswordCharacteristic_.value().c_str(),
            sizeof(singleton->wifiPassword));
    strncpy(singleton->hostName,
            singleton->hostNameCharacteristic_.value().c_str(),
            sizeof(singleton->hostName));
    strncpy(singleton->mqttBroker,
            singleton->mqttBrokerCharacteristic_.value().c_str(),
            sizeof(singleton->mqttBroker));
    strncpy(singleton->topicRoot,
            singleton->topicRootCharacteristic_.value().c_str(),
            sizeof(singleton->topicRoot));
    strncpy(singleton->location,
            singleton->locationCharacteristic_.value().c_str(),
            sizeof(singleton->location));
    singleton->sampleInterval = singleton->sampleIntervalCharacteristic_.value();
    singleton->isLockedCharacteristic_.writeValue(1);

    // Save the configuration to EEPROM
    ConfigServiceEEPROM_t flashConfig;

    // Set the config object
    flashConfig.valid = 99;
    strncpy(flashConfig.ssid, singleton->ssid, sizeof(flashConfig.ssid));
    strncpy(flashConfig.wifiPassword, singleton->wifiPassword,
            sizeof(flashConfig.wifiPassword));
    strncpy(flashConfig.hostName, singleton->hostName,
            sizeof(flashConfig.hostName));
    strncpy(flashConfig.mqttBroker, singleton->mqttBroker,
            sizeof(flashConfig.mqttBroker));
    strncpy(flashConfig.topicRoot, singleton->topicRoot,
            sizeof(flashConfig.topicRoot));
    strncpy(flashConfig.location, singleton->location,
            sizeof(flashConfig.location));
    flashConfig.sampleInterval = singleton->sampleInterval;
    strncpy(flashConfig.configurationPassword, singleton->configurationPassword,
            sizeof(flashConfig.configurationPassword));

    DEBUG_PRINTF("valid = %d\n", flashConfig.valid);
    DEBUG_PRINTF("ssid = %s\n", flashConfig.ssid);
    DEBUG_PRINTF("hostName = %s\n", flashConfig.hostName);

    flash_configuration.write(flashConfig);

    ConfigServiceEEPROM_t flashConfig2 = flash_configuration.read();

    DEBUG_PRINTF("After flash and re-read\n");
    DEBUG_PRINTF("valid = %d\n", flashConfig2.valid);
    DEBUG_PRINTF("ssid = %s\n", flashConfig2.ssid);
    DEBUG_PRINTF("hostName = %s\n", flashConfig2.hostName);

    singleton->debug_print_configuration();

  } else {
    if (singleton->lockCharacteristic_.value() ==
        singleton->configurationPassword) {
      DEBUG_PRINTF("Configuration Unlocked with password = %s\n",
                   singleton->lockCharacteristic_.value().c_str());
      singleton->isLocked = false;
      singleton->isLockedCharacteristic_.writeValue(0);
    } else {
      DEBUG_PRINTF("ERROR - incorrect password, configuation is locked!\n");
    }
  }
}

void ConfigService::debug_print_configuration() {
  DEBUG_PRINTF("--- Current Configuration ---\n");
  DEBUG_PRINTF(" SSID = %s\n", ssid);
  DEBUG_PRINTF(" wifiPasswd = %s\n", wifiPassword);
  DEBUG_PRINTF(" hostname = %s\n", hostName);
  DEBUG_PRINTF(" mqttBroker = %s\n", mqttBroker);
  DEBUG_PRINTF(" topicRoot = %s\n", topicRoot);
  DEBUG_PRINTF(" location = %s\n", location);  
  DEBUG_PRINTF(" sampleInterval = %u\n", sampleInterval);
}
